### PR TITLE
CDL 2015

### PR DESCRIPTION
**Pull request by palcu**
I will write an article about how the problem that students had to solve for the CDL and some of the questions we asked them in the interviews.

# Content

- [x] translate the [problem](http://www.infoarena.ro/problema/convertor)
- [x] write about different solutions to it
- [x] get statistics about
  - [x] number of submissions for each student who solved it
  - [x] time between submissions
  - [x] histogram of points
  - [x] histogram of failed tests
- [x] ~~get a tag cloud of the words in the form application~~ useless because it is in Romanian
- [x] write about the real life CDL interviews
  - adding the new use edge cases for key/values
  - keys are in different order and solutions like hash/BST
  - my experience w/ different students

# Steps

- [x] write what should be the content about
- [x] get data about [every submission](https://github.com/palcu/convertor/tree/master/blog_post)
- [x] write draft
- [x] proof-read
- [x] fix comments
- [x] ~~make charts transparent~~ Google Docs does not support that
- [ ] ship it tonight

### CODE REVIEW

**Comment by maria** [comment]
:+1:

**Comment by mihaimaruseac** [comment]
(y) Looking forward to it

**Comment by palcu** [comment]
@maria @mihaimaruseac @adasolcan please proof read when you have time and energy. It was fun writing this.

File: posts/2015/03/21/cdl-2015.md [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+Then I might ask them if they know a data structure where you could do lookups faster than O(log(n)). Most of them knew about hashes and gave the correct complexity solution. However, when I asked them how does a hash work, they raised their shoulders and had no idea. I then explained them that a simple implementation of a hash is a long array with a smart hash function.
+
+I must say that there were some smart students that knew what hashes were, how they worked behind the scenes and they applied them to this problem naturally without any help. For these students I asked them why would you sometimes prefer a binary search tree in place of a hash. The answer is that a BST uses lower memory. Another topic of discussion would be on how would you implement a hash function for this problem. Two students knew that using a base of 26 for the characters of the keys and then doing modulo of a big prime number would be a simple and elegant solution.
+
+All in all, I was surprised by the lack of how the students grasped the concept of a hash and applied it in the problem, but had some interesting discussions with the smart ones.
```
Comment (maria) [inline posts/2015/03/21/cdl-2015.md]: ..with some. :)

File: posts/2015/03/21/cdl-2015.md [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+After we eliminated the students who didn't have 100 points at the problem and the students who didn't complete anything on the "What project are you most proud of?" in their application form, only 60 students remained. We sent a call-to-action to the ROSEdu community and the mentors, and 8 people replied that they could help us with the interviews.
+
+Each interview took 30 minutes. For the first part we asked them questions about the non-technical applications, then asked the applicant to talk about how he approached the problem and finally we asked him the dreaded technical questions.
+
+## Technical questions
+
+We would first ask the student to present his solution. Then, we would start asking him what would happen if we modified the problem statement and he would start to have special characters inside the keys, like parentheses or colons. Then we would ask him to tell us how easily it would have been to modify the source code and support those edge cases.
+
+From what I've observed, the shorter and cleaner the student's solution was, the easier for him was to tell us a solution.
```
Comment (maria) [inline posts/2015/03/21/cdl-2015.md]: Un pic alambicata :)

File: posts/2015/03/21/cdl-2015.md [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+To select the best students, we gave the students to solve an ACM style algorithm problem on the [Infoarena](http://www.infoarena.ro/) judge. They had to code their solution in C, C++ or Java, submit it online and the judge would run it over 10 tests, checking the output and measuring the time and memory their solution took. Most of the students who tried to solve this problem were in their 1st and 2nd undergraduate years at Computer Science or Computer Engineering in Bucharest.
+
+## Problem Statement
+
+_You can read the Romanian version on [Infoarena](http://www.infoarena.ro/problema/convertor)_
+
+Ada, Calin and Andrei have got bored of learning algorithms at their University and want to learn more practical stuff. To do this, they have decided to apply at ROSEdu CDL. However, the organisers cannot separate the applicants, so they decided to give an algorithm problem for them to solve. Luckily, you don't need lots of knowledge about time complexities
```
Comment (maria) [inline posts/2015/03/21/cdl-2015.md]: have got bored -> got bored ?

Comment (maria) [inline posts/2015/03/21/cdl-2015.md]: + you forgot the `.`

File: posts/2015/03/21/cdl-2015.md [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+```csv
+id,language,usage,power,
+1,Ruby,Mainly by hipsters.,4,
+2,Python,Computer scientists and some wannabe hipsters.,2,
+3,C++,Hardcore people who love dangling pointers.,100,
+4,Haskell,A lonely dude in Massachusetts.,999999,
+```
+
+However, since we want to simulate a real life problem better, the JSON file won't be beautifully formatted. But, we guarantee it will be correct.
```
Comment (maria) [inline posts/2015/03/21/cdl-2015.md]: well formatted?

File: posts/2015/03/21/cdl-2015.md [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+```
+
+_All the complexities are assuming a comparison of strings is done in O(1)._
+
+I took about 20 interviews. Almost every student would find the O(N^2) algorithm. The solution would be to print the keys of the first object, read the second object, and for every key search it naively in the first object.
+
+From here, only half of the students would get a better solution alone. The first hint I gave was to try and see if having the keys in a certain order might help. Some of them caught the idea, sorted the keys and said that they would now use binary search to find the position of the keys in O(N*log(N)).
+
+Then I might ask them if they know a data structure where you could do lookups faster than O(log(n)). Most of them knew about hashes and gave the correct complexity solution. However, when I asked them how does a hash work, they raised their shoulders and had no idea. I then explained them that a simple implementation of a hash is a long array with a smart hash function.
+
```
Comment (maria) [inline posts/2015/03/21/cdl-2015.md]: search tree in place of a hash -> search tree rather than a hash

**Comment by maria** [comment]
review done. :)

File: posts/2015/03/21/cdl-2015.md [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+The most failed test was [number 8](https://github.com/palcu/convertor/blob/master/teste/grader_test8.in). What's special about it is that it contains a single object, and most of the students assume that after the first object ending curly brace, they will have a comma.
+
+A reason why the last 3 tests have the most failing submissions is that they are really big and most of the students preferred to use `getchar` for I/O. Doing that, you overwhelm the operating system with lots of calls. A better approach is to use a buffer.
+
+Also, I've counted the time students spent solving the problem. The average time for any student was 6 hours and 15 minutes, and the average time for a student who successfully passed all the tests was 3 hours and 45 minutes. I made the assumption that if a student submitted a problem for the first time, I added an hour to his solving time, and if the difference between 2 submissions is more than 4 hours, I'm also adding one hour for his next submission. In total, students spent 40 days of time trying to get all the tests passed.
```
Comment (mihaimaruseac) [inline posts/2015/03/21/cdl-2015.md]: Remove the "I made the assumption ...to his solving time,", replace it with "To estimate the time spent on solving the problem, I added an hour to the solving time for the first submissions,..."

File: posts/2015/03/21/cdl-2015.md (hunk 2 above)
Comment (mihaimaruseac) [inline posts/2015/03/21/cdl-2015.md]: Susțin. Aș înlocui finalul cu "the easier for him was to give us a version adapted to the new requirements"

File: posts/2015/03/21/cdl-2015.md [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+Each interview took 30 minutes. For the first part we asked them questions about the non-technical applications, then asked the applicant to talk about how he approached the problem and finally we asked him the dreaded technical questions.
+
+## Technical questions
+
+We would first ask the student to present his solution. Then, we would start asking him what would happen if we modified the problem statement and he would start to have special characters inside the keys, like parentheses or colons. Then we would ask him to tell us how easily it would have been to modify the source code and support those edge cases.
+
+From what I've observed, the shorter and cleaner the student's solution was, the easier for him was to tell us a solution.
+
+After this warming questions, we would ask him the important one: what would happen if the keys of the objects were not in the same order. For example
```
Comment (mihaimaruseac) [inline posts/2015/03/21/cdl-2015.md]: "this warming question" or "these warming questions"

File: posts/2015/03/21/cdl-2015.md (hunk 1 above)
Comment (mihaimaruseac) [inline posts/2015/03/21/cdl-2015.md]: "how students grasped", without the "the"

**Comment by mihaimaruseac** [comment]
Sorry, I just realized I spammed in here by reviewing based on commits. Solved in the future. Finished review.

**Comment by palcu** [comment]
I've updated the blog post and modified the date. I cannot make the charts transparent. I'll publish it tonight.

**Comment by mihaimaruseac** [comment]
[Published](http://techblog.rosedu.org/cdl-2015.html).
